onnxruntime/core/optimizer/selectors_actions/selector_action_transformer.cc

```diff
@@ -120,86 +151,126 @@ Status SelectorActionTransformer::MatchAndProcess(Graph& graph, const GraphViewe
-
-  const auto& actions_map = selectors_and_actions_.ActionsMap();
-  const auto actions_map_end = actions_map.cend();
-
-  for (const auto& entry : saved_actions) {
-    auto action_iter = actions_map.find(entry.action_name);
-    if (action_iter == actions_map_end) {
-      return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "Missing action ", entry.action_name, " for transformer ", Name());
+  for (auto index : graph_viewer.GetNodesInTopologicalOrder()) {
+    auto* node = graph.GetNode(index);
+    if (node == nullptr) {
+      continue;  // was removed by this transformer
     }
 
-    const std::unique_ptr<Action>& action = action_iter->second;
-
-    for (const NodesToOptimizeIndices& node_group : entry.node_groups) {
-      NodesToOptimize nodes_to_optimize{graph, node_group};
+    ORT_RETURN_IF_ERROR(Recurse(*node, modified, graph_level, logger));
 
-      // all nodes in the group are still available if IsValid returns true
-      if (nodes_to_optimize.IsValid()) {
-        ORT_RETURN_IF_ERROR(action->Run(graph, nodes_to_optimize));
-        modified = true;
-      }
+    // TODO: use GraphTransformer::GetCompatibleExecutionProviders if we need something more flexible
+    if (node->GetExecutionProviderType() == kCpuExecutionProvider) {
+      ORT_RETURN_IF_ERROR(MatchAndProcess(graph, graph_viewer, *node, modified, logger, save_context));
     }
   }
 
   return Status::OK();
 }
 
-#endif
+#endif  // !defined(ORT_MINIMAL_BUILD)
 
-Status SelectorActionTransformer::ApplyImpl(Graph& graph, bool& modified, int graph_level,
-                                            const logging::Logger& logger) const {
-  // TODO: Is there any reason to create a new GraphViewer? Do we need the different topological sort or can we
-  // just use graph.GetNodesInTopologicalOrder() and avoid the overhead of re-sorting.
-  GraphViewer graph_viewer(graph);
+#if defined(ORT_ENABLE_ORT_FORMAT_RUNTIME_GRAPH_OPTIMIZATION)
 
-  for (auto index : graph_viewer.GetNodesInTopologicalOrder()) {
-    auto* node = graph.GetNode(index);
-    if (node == nullptr) {
-      continue;  // was removed by this transformer
+static Status RegisterProducedNodesWithGraph(NodeIndex pre_action_max_num_nodes, NodeIndex post_action_max_num_nodes,
+                                             const RuntimeOptimizationRecord& record,
+                                             Graph& graph) {
+  assert(post_action_max_num_nodes >= pre_action_max_num_nodes);
+
+  const auto num_new_node_indices = post_action_max_num_nodes - pre_action_max_num_nodes;
+
+  auto produced_node_it = record.produced_nodes.begin();
+  const auto produced_nodes_end = record.produced_nodes.end();
+
+  std::unordered_map<NodeIndex, HashValue> node_index_to_kernel_def_hash{};
+
+  for (NodeIndex i = 0; i < num_new_node_indices; ++i) {
+    const NodeIndex new_node_idx = pre_action_max_num_nodes + i;
+    const auto* new_node = graph.GetNode(new_node_idx);
+
+    if (!new_node) continue;
+
```

Comment:
nit: hard to set a breakpoint on this condition being hit if it's a one-liner.

When would the new node not exist? I would have expected that would be an error.

Reply:
currently it shouldn't happen, but I thought it wouldn't be impossible for an action to add and then remove some temporary node. so here we just account for newly added nodes that still exist afterwards.

Reply:
Would be good to capture that rationale in a comment as it's not obvious why this needs to be supported.